### DISCUSSION TITLE
Proposal: use DLSS 5 for depth-guided stereo disocclusion inpainting

### DISCUSSION THREAD

**Discussion by taowen** [nagadomi/nunif#726] [Ideas]
## Motivation

IW3's depth-ordered forward warp already produces most of the geometry needed for a different hole-filling experiment:

- it warps the source x-coordinate together with the color ([`forward_warp.py`](https://github.com/nagadomi/nunif/blob/d23721f1b5f0a4c92c3ee1be013180bf298730c5/iw3/forward_warp.py#L164-L183));
- it identifies undefined/layered-hole pixels and can return per-eye masks ([same file](https://github.com/nagadomi/nunif/blob/d23721f1b5f0a4c92c3ee1be013180bf298730c5/iw3/forward_warp.py#L185-L204));
- the current lightweight path fills them by horizontal shifting, while `forward_inpaint` sends the processed mask to an image/video inpainting model ([`forward_inpaint.py`](https://github.com/nagadomi/nunif/blob/d23721f1b5f0a4c92c3ee1be013180bf298730c5/iw3/forward_inpaint.py#L18-L40)).

I would like to propose an optional experiment using **DLSS 5 Neural Rendering as a temporal, depth-guided right-eye repair stage**, not as Super Resolution or Frame Generation.

This is different from the general DLSS/VSR request in #685: the goal is specifically to reduce stereo disocclusion artifacts produced by the depth warp.

## Core idea: treat the two eyes as adjacent virtual frames

For `synthetic_view="right"`:

```text
virtual previous frame = original/left eye
virtual current frame  = warped right eye
virtual camera motion  = left-eye position -> right-eye position
```

The motion vector does not need optical-flow estimation. During the existing forward warp, retain the winning source coordinate `pL` for every destination pixel `pR`. Then generate a current-to-previous vector in right-eye coordinates:

```text
previous_pixel = current_pixel + MV
MV(pR) = pL - pR
```

For IW3's current row-wise stereo warp this is normally `float2(xL - xR, 0)`. Pixels without a winning source sample are true disocclusions: mark them in `HoleMask`, minimally prefill their color so the input texture is defined, and use zero or depth-edge-aware propagated MV there.

## Proposed pipeline

```text
RGB + estimated depth
  -> IW3 left-to-right depth-ordered forward warp
  -> right prefilled color + source-coordinate map + hole mask
  -> geometric right-to-left MV from the source map
  -> DLSS 5 Evaluate(left, Reset=1, zero MV)       # establish history
  -> DLSS 5 Evaluate(right, Reset=0, geometric MV) # virtual next frame
  -> original left + enhanced right -> SBS
```

No NVIDIA Optical Flow step is required. Each input video frame would start a fresh two-call virtual sequence (`left Reset=1`, then `right Reset=0`) so that the previous video's right eye cannot leak into the next stereo pair.

## Why this may work

I statically inspected one specific `nvngx_dlssnr.dll` build (SHA-256 `6eb209e764f39872625debd6abaf45e2bb6322f6f270f781f70c059ae30b3927`). In that build, the neural core is wired to:

- current color;
- an internal previous DLSS output;
- motion vectors;
- scalar/artistic controls.

Although the ABI parser accepts a Depth resource, that exact build does not feed Depth into the neural core. This is why the proposed integration converts IW3 depth into the warp, correspondence/MV, and hole mask **before** calling DLSS.

The build also accepts `DLSSNR.ControlMask`. Static kernel analysis indicates that it controls the final per-pixel neural-result blend:

```text
effective_blend = saturate(mask.r * blend_scale)
output = lerp(base_result, neural_result, effective_blend)
```

Therefore a dilated/feathered IW3 hole mask could concentrate the DLSS modification around disocclusions and depth edges.

Important limitation: this control mask is not an explicit inpainting-mask input to the network encoder. DLSS may make a prefilled hole look more coherent by using the aligned left-eye history and learned priors, but it cannot guarantee recovery of the true background that was never visible in the source view.

## Suggested proof of concept

1. Add an optional return value for the destination-to-source x-coordinate map in the forward-warp path.
2. Generate a full-resolution `RG16F` MV texture and `R8_UNORM` hole/control mask.
3. Implement a Windows/NVIDIA-only D3D12 NGX bridge as an optional backend. DLSS Neural Rendering uses an unpublished Feature 18 ABI, so this should remain isolated from IW3's portable PyTorch path.
4. First test only `synthetic_view="right"`, preserving the original image as the left eye.
5. Calibrate MV sign and scale using a constant-depth image and compare `MV`, `-MV`, and zero-MV outputs.
6. Verify that the Reset evaluation actually writes a history that the immediately following right-eye evaluation reads.

For evaluation, I suggest comparing:

- `forward_fill`;
- IW3 image inpainting;
- IW3 video inpainting;
- DLSS with zero MV;
- DLSS with depth-derived MV;
- DLSS with depth-derived MV plus the hole-derived control mask.

The useful criteria would be masked hole-region quality, foreground-edge stretching/halos, left-right consistency, temporal flicker, VR viewing comfort, VRAM use, and processing time.

## Integration constraints

- RTX GPU and Windows/D3D12 only for the first implementation.
- Two DLSS Neural Rendering evaluations per generated stereo pair.
- Feature 18 is not a public stable SDK ABI and may change between DLL/driver versions.
- Color encoding/HDR handling must be calibrated.
- Distributing NVIDIA DLLs or extracted kernels may have licensing implications; IW3 should not bundle them without explicit redistribution permission.

Would retaining/exporting the source-coordinate map from `depth_order_bilinear_forward_warp` fit IW3's design? If so, I can work on a small proof of concept around the `synthetic_view="right"` path first.

**Comment by taowen** [comment]
## Comparison with IW3's existing lightweight inpainting models

After reading the current implementation more closely, I think it is important to distinguish the proposed DLSS experiment from IW3's existing inpainting models.

### `LightInpaintV1` is a true mask-conditioned inpainting model

The image model receives the warped eye image and an explicit hole mask. During preprocessing it removes the pixels inside the mask:

```python
x = x * (1 - mask)
```

After 4x pixel-unshuffle, fully masked feature cells are replaced with a learned `mask_bias` before entering the spatial GMLP encoder:

```python
mask = pixel_unshuffle(mask, 4).amax(dim=1, keepdim=True) > 0.99
x = torch.where(mask, self.mask_bias, x)
```

The generated result is then composited only inside the feathered mask. See [`light_inpaint_v1.py`](https://github.com/nagadomi/nunif/blob/d23721f1b5f0a4c92c3ee1be013180bf298730c5/iw3/models/light_inpaint_v1.py#L88-L149).

This is materially different from the inspected DLSS 5 build. `DLSSNR.ControlMask` only controls the final blend between the base and neural results; it does not replace masked features or condition the network encoder. Therefore IW3's model knows explicitly that a region is missing, while DLSS only sees a minimally prefilled right-eye image plus motion-aligned history.

IW3's training data is also particularly well matched to the task. Its hole masks are generated by depth-based forward warping rather than arbitrary rectangles, and the original unwarped image is used as ground truth. This teaches the model the expected foreground/background arrangement around stereo disocclusions.

### `LightVideoInpaintV1` can use real temporal reveals

The video model processes a 12-frame sequence. Its bottleneck contains two `WindowGMLP3d` blocks with a temporal window spanning all 12 frames, interleaved with spatial GMLP blocks. See [`light_video_inpaint_v1.py`](https://github.com/nagadomi/nunif/blob/d23721f1b5f0a4c92c3ee1be013180bf298730c5/iw3/models/light_video_inpaint_v1.py#L87-L194).

The current queue uses three frames of pre-padding and three frames of post-padding, producing six frames from each full 12-frame window ([`forward_inpaint.py`](https://github.com/nagadomi/nunif/blob/d23721f1b5f0a4c92c3ee1be013180bf298730c5/iw3/forward_inpaint.py#L106-L166)). This gives the model future as well as past context, at the cost of look-ahead latency.

This is an important advantage over the proposed left-eye/right-eye virtual sequence. A true right-eye disocclusion is usually content that is not visible in the left eye at the same instant. However, camera or object motion in adjacent video frames may reveal the actual background, so the video inpainting model can sometimes use real scene information that a same-time left-eye history cannot provide.

### Practical comparison

| Property | Light image model | Light video model | Proposed DLSS 5 path |
| --- | --- | --- | --- |
| Explicit hole mask inside the network | Yes | Yes | No; final blend only |
| Temporal input | None | 12 real video frames | Left/right pair as two virtual frames |
| Explicit geometric MV | No | No | Yes, derived from the depth warp |
| Trained for IW3-style holes | Yes | Yes | Not demonstrated |
| Can use background revealed over time | No | Yes | Generally no |
| Default released weight size | about 8.7 MiB | about 8.9 MiB | about 141 MiB of embedded weights in the inspected build |
| Integration | Existing PyTorch path | Existing PyTorch path | Experimental Windows/RTX/D3D12 NGX bridge |

The lightweight models are therefore stronger baselines than simple shift filling. Model size alone should not imply that DLSS will fill holes better: the IW3 models have a much better task match, whereas DLSS has greater capacity and explicit geometric history alignment.

### Revised experiment matrix

I suggest treating DLSS as an experiment or refinement stage and comparing four pipelines:

```text
A. forward_fill
B. LightInpaintV1 / LightVideoInpaintV1
C. minimal prefill -> DLSS 5
D. LightInpaintV1 -> DLSS 5 localized refinement
```

My current expectation is:

- `LightInpaintV1` should remain the stronger task-specific single-frame hole generator.
- `LightVideoInpaintV1` should have the best opportunity to recover real background information in offline video.
- DLSS may be useful for reducing edge halos, stretched prefill textures, and locally unnatural detail after geometric alignment.
- The most plausible improvement may be the hybrid pipeline (D), rather than replacing IW3's inpainting model outright.

This comparison also provides a useful falsification test: if depth-derived MV does not measurably improve DLSS over zero MV, or if DLSS does not outperform the existing lightweight model inside the hole mask, then the extra proprietary D3D12 integration would not be justified.

**Comment by nagadomi** [comment]
If it works well and provides some useful benefits, it could be added as an option.

One concern I have, which I have not verified, is the following.
Frame interpolation and stereo warping are very similar problems, but typical frame interpolation produces an intermediate frame. There should be no problem if the position of each pixel in this intermediate frame is simply at the 0.5 position of the optical flow, or if we can specify the exact interpolation position. However, if the position is automatically adjusted based on motion vectors or their history, the warp may not be controllable as intended.

For now, I think the first thing to do is to create some test data and check whether the actual result is what you expect.

**Reply by taowen** [reply]
DLSS5 is not Frame interpolation. It is a two images to one image, image generation (a.k.a neural rendering). I "think" it works like a full image inpainting, will do some testing to verify.

**Reply by nagadomi** [reply]
Thanks for the explanation. I did not know anything about DLSS 5.
If it has a feature that can fill in areas where there is no pixel information at all, then it can be used as is.
I am not sure if the left and right views can be processed as a temporal sequence, but I think the best way is to actually try it.

**Reply by nagadomi** [reply]
Just to mention this, if there are any licensing issues with calling or investigating it, I may not be able to merge it. I will not comment on whether you use it, though.

However, that would be an uninteresting reason. I am more interested in whether it is useful in practice. If it is, it would be a good direction for future improvements.

**Reply by taowen** [reply]
managed to export the dlss5 as a pytorch model and run some experiments. https://github.com/taowen/dlss5-as-inpainting/blob/master/docs/DLSS5_STEREO_CONCLUSIONS_EN.md DLSS5 currently shows limited, material-dependent value as a local refinement stage—not a reliable general solution for stereo disocclusion filling.

**Reply by nagadomi** [reply]
I only looked at it briefly, but it seems that the control mask is just post-processing and does not affect the main network. If so, the temporary texture used to fill the holes is passed as the input context, so I think it would be difficult to achieve inpainting.
